### ISSUE TITLE
WinUI Multibinding Support

### ISSUE THREAD

**Issue by JoeTomkinson** [microsoft/microsoft-ui-xaml#8334]
# Proposal: Multibinding Support

## Summary
formerly UWP supported multibinding even as a predecessor to WPF. Multi-binding support was available for both command and converter parameters. This particular feature appears to have already been ported over to MAUI. WinUI3 still appears to lack this capability. 

## Rationale
It's useful to have multi-binding because it allows developers to create complex hierarchical data-templates for dynamic content generation whilst being able to pass bindings up and down between parent and child elements.

* Aligns WinUI with aspirations of the Project Reunion ethos, and with products such as MAUI.
* Allows for complex multi-tiered data template shenannigans.
* Should be portable from MAUI assuming they are based on a similar XAML and Class structures.

**Comment by sigmarsson** [comment]
[Yes](https://learn.microsoft.com/en-us/dotnet/maui/fundamentals/data-binding/multibinding?view=net-maui-8.0) releasing in 1.5 would be feasible ?

**Comment by GuildOfCalamity** [comment]
Is this ever coming to  WindowsAppSDK?
I'm currently using **1.5.240311000** and there is still no support for `<MultiBinding>` or `IMultiValueConverter`

**Comment by legistek** [comment]
As the original author of MultiBinding support in Xamarin Forms I would very much like to see this in WinUI. :)

The implementation was quite simple and if not for so much of WinUI being sealed or otherwise non-extensible it would be a very easy thing to implement in managed code.

**Comment by legistek** [comment]
Until this is added natively:

https://github.com/peter0302/WinUI.Redemption/

**Comment by GuildOfCalamity** [comment]
@legistek will definitely check this out, thank you... at least someone recognizes the shortcomings with WinUI

**Comment by JoeTomkinson** [comment]
> Until this is added natively:
> 
> https://github.com/peter0302/WinUI.Redemption/

That's interesting, I'll take a look at some point over the weekend, thank you for your effort on that. 

PRs from Forks ok if there's any additional value I can add in there?

**Comment by legistek** [comment]
@GuildOfCalamity yeah, finally being able to see the code explains a lot though. The entire projection system from native to managed is batty and I'm convinced explains at least 50% of the frustration. One also wonders how many seemingly senseless restrictions and counterproductive performance tweaks we're still living with that were just done for the Surface RT. Visually the framework is amazing but I wish they'd have burned the RT codebase and started from scratch for Windows 10 with something more sensibly balanced between native and managed like WPF was. Alas. But anyway now that it's decoupled from the OS they can be more liberal with changes that previously would've required an OS update, so I really hope they start tackling more of these longstanding issues now. Maybe my little project can help things along.

@JoeTomkinson please by all means.

**Comment by GuildOfCalamity** [comment]
@JoeTomkinson What baffles me is the fact that this was already working in UWP... why remove it?

**Comment by legistek** [comment]
I don't think MultiBinding ever existed in Metro or UWP. MultiBinding works with a MAUI app running atop UWP (or any other platform current or future), but it's still implemented at the MAUI layer. So if you're coming from MAUI (or WPF) to WinUI it looks like it was taken out but it wasn't ever there to begin with.

Incidentally WPF's implementation is *completely* different from the MAUI implementation and is probably 100x the lines of code. 

I always thought my solution in XF was on the hackish-side, just as this new one definitely is, but then again if you look at the guts of MAUI/XF a whole lot of it fits that description. I imagine a native C++ implementation that used the WinUI internals would take a different approach, but who knows.

**Comment by GuildOfCalamity** [comment]
@legistek Well, at any rate, your code works.
I had an issue where I needed to pass two date objects in an ObservableCollection source for Multibinding color switching.
One date is the creation time and the other is the updated time; the further the dates get apart from one another the color then changes to match their difference.

**Comment by contextfree** [comment]
> I don't think MultiBinding ever existed in Metro or UWP

or Silverlight (which was the codebase that Metro/UWP/WinUI3 was originally forked from)

**Comment by legistek** [comment]
Wait, was Silverlight written in native C++? I always assumed it was the same codebase as WPF. But it explains A LOT if that was the foundation for Metro.

(Sidenote - If Silverlight was always C++ why in heaven's name didn't they port it to WASM/Mono instead of killing it??).

**Comment by contextfree** [comment]
Yeah, Silverlight was a rewrite of the WPF core layout engine into C++, with a C# managed API layer on top. (Metro/WinRT XAML kept the layout engine and replaced that API layer with a WinRT one.)

**Comment by GuildOfCalamity** [comment]
If anyone needs to refer to my test of @legistek multibinding support:
[TabApp](https://github.com/GuildOfCalamity/TabApp/blob/main/Views/Tab2Page.xaml)

**Comment by legistek** [comment]
@GuildOfCalamity oh well MultiBinding as a whole is ultimately threadsafe because I `lock` `_proxyProps` it before using it, which is basically what any `Concurrent` collection would do. That was just a note so no one else forgot to lock it before using it. In any case I wanted the underlying collection to just be an array because of how infrequently items will be added vs. retrieved. Also items need to be easily retrieved by index for the whole "fungible AttachedProperty's" thing to work.

**Comment by Katerlad** [comment]
Hello! One year later in the future, it appears Multibinding is still not possible? I have an idea for an app for windows, started down the path of creating a hexagon user control, realized I need a Converter to calculate the polygons points, to calculate the points properly I need to pass a radius, the orientation of the hexagon, and the stroke thickness of its border... 

A normal IValueConverter I can only bind one Value, and one custom parameter.... 

This is very annoying that someone new to this ecosystem finds out this is something that has been supported in all of the old Windows Development platforms, but it is not supported or removed in the new platform (WinUI3). 

This is my first 16 work hour experience with WinUI3 so far...... Makes me want to switch to something else like React Native Desktop.

**Comment by greatvova** [comment]
Hello. 
I've complex app and i've decided to add windows support to it. 
Unfortunately my app uses 19 multi bindings and there is no way they can be replaced with anything without the great loss of performance.
When this feature will be added for WinUI projects?
